stdlib/fs.gr

```diff
@@ -0,0 +1,1049 @@
+  modifiedTimestamp: Number,
+  /** Last file status change timestamp in nanoseconds */
+  changedTimestamp: Number,
+}
+
+/**
+ * Represents information about an item in a directory.
+ */
+provide record DirectoryEntry {
+  name: String,
+  fileType: FileType,
+}
+
+/**
+ * The type of removal operation to perform when calling `remove`.
+ */
+provide enum RemoveMode {
+  RemoveFile,
+  RemoveEmptyDirectory,
+  RemoveRecursive,
+}
+
+/**
+ * The type of copy operation to perform when calling `copy`.
+ */
+provide enum CopyMode {
+  CopyFile,
+  CopyRecursive,
+}
+
+/**
+ * The type of write operation to perform when calling `writeFile`.
+ */
+provide enum WriteMode {
+  Truncate,
+  Append,
+}
+
+enum OpenMode {
+  Unlink,
+  RmDir,
+  RmrfDir,
+  CopySourceBase,
+  CopyTargetBase,
+  CopySource,
+  CopyTarget,
+  RenameSource,
+  RenameTarget,
+  ReadDir,
+  MakeDir,
+  MakeSymlink,
+  ReadLink,
+  Stats,
+  Read,
+  WriteFile,
+  AppendFile,
+}
+
+let fileResult = result => {
+  match (result) {
+    Ok(ok) => Ok(ok),
+    Err(Wasi.SystemError(err)) => Err(errnoToFileError(err)),
+    _ => fail "Impossible: non-wasi error",
+  }
+}
+
+let openInfo = openMode => {
+  match (openMode) {
+    Unlink => ([File.Directory: File.OpenFlag], [File.PathUnlinkFile], [], []),
+    RmDir => ([File.Directory], [File.PathRemoveDirectory], [], []),
+    RmrfDir => {
+      let rights = [
+        File.PathOpen,
+        File.PathFilestats,
+        File.PathRemoveDirectory,
+        File.PathUnlinkFile,
+        File.FdReaddir,
+      ]
+      ([File.Directory], rights, rights, [])
+    },
+    CopySourceBase => {
+      let rightsInheriting = [
+        File.PathOpen,
+        File.FdReaddir,
+        File.PathFilestats,
+        File.PathReadlink,
+      ]
+      (
+        [File.Directory],
+        rightsInheriting,
+        [File.FdRead, ...rightsInheriting],
+        [],
+      )
+    },
+    CopyTargetBase => {
+      let rightsInheriting = [
+        File.PathOpen,
+        File.PathCreateDirectory,
+        File.PathCreateFile,
+        File.PathSetSize,
+        File.PathSymlink,
+      ]
+      (
+        [File.Directory],
+        rightsInheriting,
+        [File.FdWrite, ...rightsInheriting],
+        [],
+      )
+    },
+    CopySource => ([], [File.FdRead], [], []),
+    CopyTarget => ([File.Create, File.Truncate], [File.FdWrite], [], []),
+    RenameSource => ([File.Directory], [File.PathRenameSource], [], []),
+    RenameTarget =>
+      ([File.Directory], [File.PathRenameTarget, File.PathFilestats], [], []),
+    ReadDir => ([File.Directory], [File.FdReaddir], [], []),
+    MakeDir => ([File.Directory], [File.PathCreateDirectory], [], []),
+    MakeSymlink => ([File.Directory], [File.PathSymlink], [], []),
+    ReadLink =>
+      ([File.Directory], [File.PathFilestats, File.PathReadlink], [], []),
+    Stats => ([File.Directory], [File.PathFilestats], [], []),
+    Read => ([], [File.FdRead, File.FdFilestats], [], [File.Rsync]),
```

Comment:
The `File.Rsync` seems to be causing `NotSupportedOrOperationNotSupportedOnSocket` to be thrown when running with wasmtime.

Reply:
It'll really depend on the runtime implementation.

Reply:
> It'll really depend on the runtime implementation.

I understand that but I still think we could improve the experience a little by having a way to disable the flag maybe something like `Fs.disableRsync`, or providing a way for users to override stuff like this, the workaround is directly using `Wasi/File` which doesn't seem ideal especially given how popular wasmtime is.

Reply:
Added a flag as a default argument to the read and write functions